amr-wind/wind_energy/ABLBoundaryPlane.cpp

```diff
@@ -0,0 +1,624 @@
+    : m_time(sim.time()), m_repo(sim.repo()), m_mesh(sim.mesh())
+{
+    amrex::ParmParse pp("ABL");
+    pp.query("bndry_write_frequency", m_write_frequency);
+    pp.queryarr("bndry_output_planes", m_out_planes);
+    pp.query("bndry_output_start_time", m_out_start_time);
+    pp.queryarr("bndry_output_var_names", m_out_var_names);
+    pp.queryarr("bndry_input_var_names", m_in_var_names);
+    pp.query("bndry_output_file", m_outfile);
+    pp.query("bndry_input_file", m_infile);
+
+    for (const auto& plane : m_out_planes) {
+        amrex::Vector<std::string> valid_planes{{"xlo", "ylo"}};
+
+        if (std::find(valid_planes.begin(), valid_planes.end(), plane) ==
+            m_out_planes.end()) {
+            throw std::runtime_error(
+                "Requested plane does not exist: pick one of [xlo, ylo]");
+        }
+    }
+
+    for (const auto& fname : m_out_var_names) {
+        if (m_repo.field_exists(fname)) {
+            auto& fld = m_repo.get_field(fname);
+            m_out_fields.emplace_back(&fld);
+        } else {
+            amrex::Print() << "  Invalid output variable requested: " << fname
+                           << std::endl;
+        }
+    }
+
+    for (const auto& fname : m_in_var_names) {
+        if (m_repo.field_exists(fname)) {
+            auto& fld = m_repo.get_field(fname);
+            fld.register_fill_patch_op<ABLFillInflow>(m_mesh, m_time, *this);
+            m_in_fields.emplace_back(&fld);
+        } else {
+            amrex::Print() << "  Invalid input variable requested: " << fname
+                           << std::endl;
+        }
+    }
+}
+
+void ABLBoundaryPlane::write_header()
+{
+    if (m_outfile == "") return;
+
+    amrex::Print() << "Creating output NetCDF file: " << m_outfile << std::endl;
+
+    auto ncf = ncutils::NCFile::create_par(
+        m_outfile, NC_CLOBBER | NC_NETCDF4 | NC_MPIIO,
+        amrex::ParallelContext::CommunicatorSub(), MPI_INFO_NULL);
+
+    ncf.enter_def_mode();
+    ncf.def_dim("sdim", 1);
+    ncf.def_dim("pdim", 2);
+    ncf.def_dim("vdim", 3);
+    ncf.def_dim("nt", NC_UNLIMITED);
+    ncf.def_var("time", NC_DOUBLE, {"nt"});
+
+    const int nlevels = m_repo.num_active_levels();
+    for (amrex::OrientationIter oit; oit; ++oit) {
+        auto ori = oit();
+        const std::string plane = m_plane_names[ori];
+
+        if (std::find(m_out_planes.begin(), m_out_planes.end(), plane) ==
+            m_out_planes.end())
+            continue;
+
+        auto plane_grp = ncf.def_group(plane);
+
+        const int normal = ori.coordDir();
+        auto v_normal = plane_grp.def_scalar("normal", NC_INT);
+        v_normal.put(&normal);
+
+        const int face_dir = ori.faceDir();
+        auto v_face = plane_grp.def_scalar("side", NC_INT);
+        v_face.put(&face_dir);
+
+        const amrex::GpuArray<int, 2> perp = get_perpendicular_idx(normal);
+        auto v_perp = plane_grp.def_var("perpendicular", NC_INT, {"pdim"});
+        v_perp.put(perp.data());
```

Comment:
This might not work on GPUs if AMReX switches from UVM to device memory.

Reply:
fixed